public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts

```diff
@@ -0,0 +1,156 @@
+import { createAction, createReducer } from '@reduxjs/toolkit';
+
+import { DataQuery, RelativeTimeRange, getDefaultRelativeTimeRange } from '@grafana/data';
+import { getNextRefIdChar } from 'app/core/utils/query';
+import {
+  dataSource as expressionDatasource,
+  ExpressionDatasourceUID,
+} from 'app/features/expressions/ExpressionDatasource';
+import { isExpressionQuery } from 'app/features/expressions/guards';
+import { ExpressionQuery, ExpressionQueryType } from 'app/features/expressions/types';
+import { defaultCondition } from 'app/features/expressions/utils/expressionTypes';
+import { AlertQuery } from 'app/types/unified-alerting-dto';
+
+import { getDefaultOrFirstCompatibleDataSource } from '../../../utils/datasource';
+import { queriesWithUpdatedReferences, refIdExists } from '../util';
+
+interface State {
+  queries: AlertQuery[];
+}
+
+const initialState: State = {
+  queries: [],
+};
+
+export const addNewDataQuery = createAction('addNewDataQuery');
+export const setDataQueries = createAction<AlertQuery[]>('setDataQueries');
+
+export const addNewExpression = createAction('addNewExpression');
+export const removeExpression = createAction<string>('removeExpression');
+export const updateExpression = createAction<ExpressionQuery>('updateExpression');
+export const updateExpressionRefId = createAction<{ oldRefId: string; newRefId: string }>('updateExpressionRefId');
+export const rewireExpressions = createAction<{ oldRefId: string; newRefId: string }>('rewireExpressions');
+export const updateExpressionType = createAction<{ refId: string; type: ExpressionQueryType }>('updateExpressionType');
+
+export const queriesAndExpressionsReducer = createReducer(initialState, (builder) => {
```

Comment:
💬 I'm using `createReducer` here from `@redux/toolkit`, which works with React's `useReducer` but comes with support for `Immer` out-of-the-box when we add our reduce functions.

Each reduce function receives `state` which is a "draft" that you can mutate, Immer will take care of computing the correct delta and apply it immutably.

https://redux-toolkit.js.org/usage/immer-reducers

Reply:
Nice idea! I like the Immer's approach, I find it less error-prone

Reply:
Great job with the reducer for queries/expressions handling!